### PR TITLE
Bootstrap FC optimization: Interfaces for state build and state transition flows

### PR DESCRIPTION
**Pull request by piyujai**

### CODE REVIEW

File: ambry-api/src/main/java/com/github/ambry/config/ClusterMapConfig.java [hunk 1]
```diff
@@ -372,6 +372,11 @@ public class ClusterMapConfig {
   @Default("false")
   public final boolean clusterMapIgnoreDownwardStateTransition;
 
+  public static final String ENABLE_FILE_COPY_FOR_BOOTSTRAP = "clustermap.enable.file.copy.for.bootstrap";
+  @Config(ENABLE_FILE_COPY_FOR_BOOTSTRAP)
+  @Default("false")
```
Comment (piyujai) [inline ambry-api/src/main/java/com/github/ambry/config/ClusterMapConfig.java:377]: Merge into single storeConfig.

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java [hunk 2]
```diff
@@ -1002,6 +1009,33 @@ public void onPartitionBecomeDroppedFromOffline(String partitionName) {
...
+      } catch (StoreException e) {
+        logger.error("Failed to start new added store for partition id {}", replica.getPartitionId(), e);
+      }
+      DiskManager diskManager = partitionToDiskManager.get(replica.getPartitionId());
+      if (diskManager == null){
+        logger.error("DiskManager not found for partition {} in partitionToDiskManager", partitionName);
+        throw new StateTransitionException(
+            "Existing Disk manager not found for replica " + partitionName + " is not found in clustermap for " + currentNode, DiskManagerNotFoundForFileCopyStateBuild);
+      }
+      diskManager.addBlobStoreToCompactionManager((BlobStore) store);
```
Comment (piyujai) [inline ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:1036]: store.start() and diskManager.addBlobStoreToCompactionManager((BlobStore) store) can be moved to file copy manager class.

File: ambry-api/src/main/java/com/github/ambry/config/ClusterMapConfig.java [hunk 3]
```diff
@@ -372,6 +372,11 @@ public class ClusterMapConfig {
...
+  public final boolean enableFileCopyForBootstrap;
```
Comment (DevenAhluwalia) [inline ambry-api/src/main/java/com/github/ambry/config/ClusterMapConfig.java:378]: Keeping the nomenclature in this file consistent, the name of this key would be 
`clusterMapEnableFileCopyForBootstrap`

File: ambry-api/src/main/java/com/github/ambry/config/StoreConfig.java [hunk 4]
```diff
@@ -674,6 +674,13 @@ public class StoreConfig {
   public final boolean storeBlockStaleBlobStoreToStart;
   public final static String storeBlockStaleBlobStoreToStartName = "store.block.stale.blob.store.to.start";
 
+
+  @Config(ENABLE_FILE_COPY_FOR_BOOTSTRAP)
+  @Default("false")
+  public final boolean enableFileCopyForBootstrap;
+  public static final String ENABLE_FILE_COPY_FOR_BOOTSTRAP = "clustermap.enable.file.copy.for.bootstrap";
```
Comment (DevenAhluwalia) [inline ambry-api/src/main/java/com/github/ambry/config/StoreConfig.java:681]: Either this or the the key defined in ClusterMapConfig needs to be removed.
Further, 
1. Key in this file will have `store` as prefix
2. Keeping the nomenclature in this file consistent, the key name would be `storeEnableFileCopyForBootstrap`

File: ambry-store/src/main/java/com/github/ambry/store/DiskManager.java [hunk 5]
```diff
@@ -447,12 +455,22 @@ boolean addBlobStore(ReplicaId replica) {
...
             replicaStatusDelegates, time, accountService, null, indexPersistScheduler);
-        store.start();
+        // For Filecopy, we do not init the BlobStore since new log/index isn't required, the files will be directly
+        // copied from remote node.
+        if (!storeConfig.enableFileCopyForBootstrap) {
+          store.start();
+        }
+
         // collect store segment requirements and add into DiskSpaceAllocator
         List<DiskSpaceRequirements> storeRequirements = Collections.singletonList(store.getDiskSpaceRequirements());
```
Comment (DevenAhluwalia) [inline ambry-store/src/main/java/com/github/ambry/store/DiskManager.java:465]: If `store.start()` hasn't been invoked, in the normal Replication protocol, would `store.getDiskSpaceRequirements()` work as expected?

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java [hunk 6]
```diff
@@ -755,6 +760,8 @@ public void onPartitionBecomeBootstrapFromOffline(String partitionName) {
           }
         } while (!replicaAdded);
 
+
```
Comment (DevenAhluwalia) [inline ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:763]: nit: remove unnecessary new lines.

Comment (DevenAhluwalia) [reply ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:763]: Applicable throughout the diff wherever applicable.

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java (hunk 2 above)
Comment (DevenAhluwalia) [reply ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:1036]: Can you elaborate as to why would that be advantageous?

**Review by DevenAhluwalia — CHANGES_REQUESTED** [review]

**Comment by DevenAhluwalia** [comment]
This PR needs to be opened with target branch `bootstrap-optimisation-lld-flow`

File: ambry-api/src/main/java/com/github/ambry/config/StoreConfig.java (hunk 4 above)
Comment (piyujai) [reply ambry-api/src/main/java/com/github/ambry/config/StoreConfig.java:681]: Noted.

File: ambry-file-transfer/src/main/java/com/github/ambry/FileCopyManager.java [hunk 7]
```diff
@@ -41,7 +41,8 @@ class PartitionStateChangeListenerImpl implements PartitionStateChangeListener {
 
     @Override
     public void onPartitionBecomeBootstrapFromOffline(String partitionName) {
-
+      // StateBuilding at the end of FCM's Offline->Bootstrap transition
+      storageManagerListener.buildStateForFileCopy(partitionName);
```
Comment (piyujai) [inline ambry-file-transfer/src/main/java/com/github/ambry/FileCopyManager.java:45]: move to storeManager

**Review by DevenAhluwalia — APPROVED** [review]

**Pull request merged**